### PR TITLE
Booking Locations API

### PR DESCRIPTION
**Pull request by benlovell**
Adds a booking locations API endpoint to facilitate the centralised bookings client functionality.

Details are in the individual commits.

### Request ###

`GET /api/v1/booking_locations/:uid`

### Response ###

```
{
  "uid": "9d7c72fc-0c74-4418-8099-e1a4e704cb01",
  "name": "Romford CAB",
  "address": "3 Some Road, Place, Romford, Essex, RM107BD",
  "locations": [
    // nested locations when present
  ],
  "guiders": [
    {
      "id": 1,
      "name": "Ben Lovell",
      "email": "benjamin.lovell@gmail.com"
    }
  ],
  "slots": [
    {
      "date": "2016-06-13",
      "start": "0900",
      "end": "1300"
    },
    {
      "date": "2016-06-13",
      "start": "1300",
      "end": "1700"
    }

    // further slots truncated
  ]
}
```

### CODE REVIEW

File: app/models/slot.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    slot_dates.map do |date|
+      [
+        new(date.iso8601, '0900', '1300'),
+        new(date.iso8601, '1300', '1700')
+      ]
+    end.flatten
+  end
+
+  def self.slot_dates
+    (Time.zone.today + 2.days..6.weeks.from_now).reject do |date|
```
Comment (benlovell) [inline app/models/slot.rb]: I'm putting these in constants.

File: spec/features/booking_locations_api_spec.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def then_the_response_is_ok
+    expect(page.status_code).to eq(200)
+  end
+
+  def and_the_booking_location_is_returned_as_json
+    expect(JSON.parse(page.body)).to include('uid' => @booking_location.uid)
+  end
+end
```
Comment (benbarnett) [inline spec/features/booking_locations_api_spec.rb]: Is it worth testing the unhappy path for an non-existent location too? The client will need to know what response to expect – probably just a 404.

Also, I'm not sure if this belongs in this PR, but we'll need a way of telling the client whether online booking is enabled for this booking location... perhaps the same 200/404 responses would suffice for this too?

**Comment by benbarnett** [comment]
Wondering if it's worth adding a `requested_uid` to the response object too... let me tell you for why.

* Customer selects 'book online' for a child location `child_uid`
* API call made for `child_uid`
* Response returned for `parent_uid` with nested `child_uid` in `locations`
* Client would need to walk `locations` array until it finds `child_uid` it originally asked for in order to display the correct location title and meta data to the customer

This logic is fine, but the client needs to keep track of the `child_uid` itself. If the `requested_uid` came back with the response object, it could just use this to pluck the original location straight from the `locations` array, and be sure that the locations being talked about are the right ones.

**Comment by benbarnett** [comment]
As discussed we'll always have a reference to the requested_uid in the request itself, and can't think of a time you would ever not use the specific location to make the call – so adding it to the response isn't worth it.

File: spec/features/booking_locations_api_spec.rb (hunk 2 above)
Comment (benlovell) [inline spec/features/booking_locations_api_spec.rb]: Good point. I'll add some tests and another path that 404s in this instance.

Comment (benlovell) [inline spec/features/booking_locations_api_spec.rb]: Addressed in 8e05cbf.

**Comment by benbarnett** [comment]
🛳🛳🛳🛳🛳🛳

**Pull request merged**